### PR TITLE
Extend solidity-frontend to support Method Overload, Inheritance and more return types for functions calls.

### PR DESCRIPTION
**Pull request by Sleepytt1210**
- Unique id is create for methods' symbol to avoid name clashes. This allows method overloading to be correctly converted.
The symbol.id of the methods are formalised as `id = “c:@C@” + contract_name + “@F@” + function_name + “#” + function_id`.

- For method calling, the function declaration node is retrieved by using the *"referencedDeclaration"* field which returned the id of the original function declaration node. 

- Added support for more return types by using get_type_descriptions to parse the *"typeDescriptions"* fields.

- Also fixed a bug where integer literals were always returned as bit vector type of `width: 32` for ReturnStatement.

Relevant test cases are added in `regression/esbmc-solidity/` to test the correctness.

### CODE REVIEW

File: regression/esbmc-solidity/return_4/contract.sol [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity >=0.5.0;
+
+// Ensure functions can return different type of value.
+contract A {
```
Comment (kunjsong01) [inline regression/esbmc-solidity/return_4/contract.sol:5]: What's the purpose of this test case?

File: regression/esbmc-solidity/polymorphism_1/polymorphism.sol [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+contract B {
+    function poly() public pure returns (uint8) {
+        return 42;
+    }
+}
+
+contract C {
+    function test_poly() public {
+        A contract_A = new A();
```
Comment (kunjsong01) [inline regression/esbmc-solidity/polymorphism_1/polymorphism.sol]: This test case(TC) doesn't cover polymorphism. Please update the name for this TC.

File: regression/esbmc-solidity/polymorphism_4/polymorphism.sol [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity >=0.5.0;
+
+// Methods overloadin: contract_A.poly(-2) should call the poly(int8) 
+// while contract_A.poly(300) should call poly(uint16).
+contract A {
+
+    function poly(uint16 b) public pure returns (uint16) {
```
Comment (kunjsong01) [inline regression/esbmc-solidity/polymorphism_4/polymorphism.sol]: This TC tests function overloading. Please update the TC name.

File: regression/esbmc-solidity/polymorphism_1/polymorphism.sol (hunk 2 above)
Comment (kunjsong01) [reply regression/esbmc-solidity/polymorphism_1/polymorphism.sol]: also for other TCs named `polymorphism_*`

File: src/solidity-frontend/solidity_convert.h [hunk 4]
```diff
@@ -186,6 +186,248 @@ class solidity_convertert
   bool get_elementary_type_name_int(
     SolidityGrammar::ElementaryTypeNameT &type,
     typet &out);
+  
+  /*
```
Comment (kunjsong01) [inline src/solidity-frontend/solidity_convert.h]: There's no implementation for the method declarations below. please remove if not needed.

**Review by kunjsong01 — CHANGES_REQUESTED** [review]
1. This PR adds support for basic inheritance rather than polymorphism. Please update the PR title/description, as well as the names for the test cases added. 
2. For a passing TC, we usually add failing TC (see test cases in `src/regression/esbmc` - `to_union_0` and `to_union_0_fail`)

File: regression/esbmc-solidity/polymorphism_4/polymorphism.sol [hunk 5]
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity >=0.5.0;
+
+// Methods overloadin: contract_A.poly(-2) should call the poly(int8)
```
Comment (kunjsong01) [inline regression/esbmc-solidity/polymorphism_4/polymorphism.sol]: typo - overloading

File: regression/esbmc-solidity/polymorphism_1/polymorphism.sol [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        B contract_B = new B();
+        assert(contract_A.poly() == 21);
```
Comment (ChenfengWei0) [inline regression/esbmc-solidity/polymorphism_1/polymorphism.sol]: May I ask if the verification fails if we change this assertion to `assert(contract_A.poly() == 42);`

**Comment by ChenfengWei0** [comment]
Thank you very much for your work. Here is my comments:
-  You might need to consider the implicit constructor. 
    - In current ESBMC-Solidity, if there is no constructor function, we will manually add a default constructor by `add_implicit_constructor()`. 
- So you might need to change the `id` format in:
```c++
 bool solidity_convertert::add_implicit_constructor()
{
  std::string name, id;
  name = id = current_contractName;     // set the idstring format 
```
and in any place where a function symbol is looked up by calling `context.find_symbol(id)`, e.g
```c++
bool solidity_convertert::get_implicit_ctor_call(
  const int ref_decl_id,
  exprt &new_expr)
{
  std::string name, id;
  if(get_contract_name(ref_decl_id, name))
    return true;
  id = name;      // set the id format 
  if(context.find_symbol(id) == nullptr)
    return true;
```

File: src/solidity-frontend/solidity_convert.cpp [hunk 7]
```diff
@@ -1179,12 +1186,24 @@ bool solidity_convertert::get_expr(
...
+    const int contract_id = caller_expr_json["scope"].get<std::uint16_t>();
+
+    std::string ref_contract_name;
+    get_contract_name(contract_id, ref_contract_name);
+
     // obtain type info from symbol table
     std::string name, id;
     name = callee_expr_json["memberName"];
-    id = name;
+    id = "c:@C@" + ref_contract_name + "@F@" + name + "#" + i2string(caller_id);
```
Comment (ChenfengWei0) [inline src/solidity-frontend/solidity_convert.cpp]: Is it possible to obtain the `id` uniformly by `get_function_definition_name()`?

File: src/solidity-frontend/solidity_convert.cpp [hunk 8]
```diff
@@ -1179,12 +1186,24 @@ bool solidity_convertert::get_expr(
   case SolidityGrammar::ExpressionT::MemberCallClass:
   {
     assert(expr.contains("expression"));
+    expr.dump(4, ' ', false, nlohmann::detail::error_handler_t::strict).c_str();
```
Comment (ChenfengWei0) [inline src/solidity-frontend/solidity_convert.cpp]: May I ask if you can add some comments here?

File: src/solidity-frontend/solidity_convert.cpp [hunk 9]
```diff
@@ -471,7 +476,8 @@ bool solidity_convertert::get_access_from_decl(
 }
 
 bool solidity_convertert::get_function_definition(
-  const nlohmann::json &ast_node)
+  const nlohmann::json &ast_node,
+  exprt &new_expr)
```
Comment (ChenfengWei0) [inline src/solidity-frontend/solidity_convert.cpp]: Why do we need the `new_expr` here?

File: regression/esbmc-solidity/return_4/contract.sol (hunk 1 above)
Comment (Sleepytt1210) [reply regression/esbmc-solidity/return_4/contract.sol:5]: My mistake, this should test for member calls of other integer return types. I will add the member calls assertion below.

File: regression/esbmc-solidity/polymorphism_1/polymorphism.sol (hunk 2 above)
Comment (Sleepytt1210) [reply regression/esbmc-solidity/polymorphism_1/polymorphism.sol]: Will do.

File: src/solidity-frontend/solidity_convert.h (hunk 4 above)
Comment (Sleepytt1210) [reply src/solidity-frontend/solidity_convert.h]: No problem.

**Review by ChenfengWei0 — CHANGES_REQUESTED** [review]

File: regression/esbmc-solidity/polymorphism_1/polymorphism.sol (hunk 6 above)
Comment (Sleepytt1210) [reply regression/esbmc-solidity/polymorphism_1/polymorphism.sol]: Yes, the verification failed with counterexample given:
```
State 1 file polymorphism.sol line 23 function test_poly thread 0
----------------------------------------------------
Violated property:
  file polymorphism.sol line 23 function test_poly
  assertion
  (signed int)return_value$_poly$3 == 42


VERIFICATION FAILED
```

**Comment by lucasccordeiro** [comment]
Can you add this case as a regression test?Sent from my iPhoneOn 12 May 2023, at 22:25, Sleepytt1210 ***@***.***> wrote:﻿
@Sleepytt1210 commented on this pull request.



In regression/esbmc-solidity/polymorphism_1/polymorphism.sol:
> +    function poly() public pure returns (uint8) {
+        return 21;
+    }
+}
+
+contract B {
+    function poly() public pure returns (uint8) {
+        return 42;
+    }
+}
+
+contract C {
+    function test_poly() public {
+        A contract_A = new A();
+        B contract_B = new B();
+        assert(contract_A.poly() == 21);

Yes, the verification failed with counterexample given:
State 1 file polymorphism.sol line 23 function test_poly thread 0
----------------------------------------------------
Violated property:
  file polymorphism.sol line 23 function test_poly
  assertion
  (signed int)return_value$_poly$3 == 42


VERIFICATION FAILED

File: src/solidity-frontend/solidity_convert.h (hunk 4 above)
Comment (Sleepytt1210) [reply src/solidity-frontend/solidity_convert.h]: Hi, I think the implementation can be found at solidity_convert.cpp:2074. It's used in the get_elementary_type() function. So I'll keep it.

File: src/solidity-frontend/solidity_convert.cpp (hunk 7 above)
Comment (Sleepytt1210) [reply src/solidity-frontend/solidity_convert.cpp]: Yes, I will make corresponding changes.

File: src/solidity-frontend/solidity_convert.cpp (hunk 8 above)
Comment (Sleepytt1210) [reply src/solidity-frontend/solidity_convert.cpp]: Sorry about this, I think I need to remove this. I wrote this to bypass the inline function call error during the gdb debugging session.

File: src/solidity-frontend/solidity_convert.cpp (hunk 9 above)
Comment (Sleepytt1210) [reply src/solidity-frontend/solidity_convert.cpp]: It was a dummy expression, I will remove it.

File: src/solidity-frontend/solidity_convert.h (hunk 4 above)
Comment (ChenfengWei0) [reply src/solidity-frontend/solidity_convert.h]: I think he means the code changes in lines 189 to 430.

Comment (Sleepytt1210) [reply src/solidity-frontend/solidity_convert.h]: Noted, thanks!

**Comment by ChenfengWei0** [comment]
> Thank you very much for your work. Here is my comments:
> 
>     - You might need to consider the implicit constructor.
>       
>       - In current ESBMC-Solidity, if there is no constructor function, we will manually add a default constructor by `add_implicit_constructor()`.
> 
>     - So you might need to change the `id` format in:
> 
> ```c++
>  bool solidity_convertert::add_implicit_constructor()
> {
>   std::string name, id;
>   name = id = current_contractName;     // set the idstring format 
> ```
> 
> and in any place where a function symbol is looked up by calling `context.find_symbol(id)`, e.g
> 
> ```c++
> bool solidity_convertert::get_implicit_ctor_call(
>   const int ref_decl_id,
>   exprt &new_expr)
> {
>   std::string name, id;
>   if(get_contract_name(ref_decl_id, name))
>     return true;
>   id = name;      // set the id format 
>   if(context.find_symbol(id) == nullptr)
>     return true;
> ```

---

May I ask if you have a plan to implement this? It could be another PR though. 

Also, you may want to resolve the build error. The PR can only be merged if all checks are passed. 
```sh
In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
../src/solidity-frontend/solidity_convert.h: In constructor 'solidity_convertert::solidity_convertert(contextt&, nlohmann::json&, const string&, const string&)':
../src/solidity-frontend/solidity_convert.h:48:15: error: 'solidity_convertert::current_contractName' will be initialized after [-Werror=reorder]
   48 |   std::string current_contractName;
      |               ^~~~~~~~~~~~~~~~~~~~
In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
../src/solidity-frontend/solidity_convert.h:38:16: error:   'unsigned int solidity_convertert::current_scope_var_num' [-Werror=reorder]
   38 |   unsigned int current_scope_var_num;
      |                ^~~~~~~~~~~~~~~~~~~~~
../src/solidity-frontend/solidity_convert.cpp:15:1: error:   when initialized here [-Werror=reorder]
   15 | solidity_convertert::solidity_convertert(
      | ^~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

**Comment by Sleepytt1210** [comment]
> > Thank you very much for your work. Here is my comments:
> > ```
> > - You might need to consider the implicit constructor.
> >   
> >   - In current ESBMC-Solidity, if there is no constructor function, we will manually add a default constructor by `add_implicit_constructor()`.
> > 
> > - So you might need to change the `id` format in:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```c++
> >  bool solidity_convertert::add_implicit_constructor()
> > {
> >   std::string name, id;
> >   name = id = current_contractName;     // set the idstring format 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and in any place where a function symbol is looked up by calling `context.find_symbol(id)`, e.g
> > ```c++
> > bool solidity_convertert::get_implicit_ctor_call(
> >   const int ref_decl_id,
> >   exprt &new_expr)
> > {
> >   std::string name, id;
> >   if(get_contract_name(ref_decl_id, name))
> >     return true;
> >   id = name;      // set the id format 
> >   if(context.find_symbol(id) == nullptr)
> >     return true;
> > ```
> 
> May I ask if you have a plan to implement this? It could be another PR though.
> 
> Also, you may want to resolve the build error. The PR can only be merged if all checks are passed.
> 
> ```shell
> In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
> ../src/solidity-frontend/solidity_convert.h: In constructor 'solidity_convertert::solidity_convertert(contextt&, nlohmann::json&, const string&, const string&)':
> ../src/solidity-frontend/solidity_convert.h:48:15: error: 'solidity_convertert::current_contractName' will be initialized after [-Werror=reorder]
>    48 |   std::string current_contractName;
>       |               ^~~~~~~~~~~~~~~~~~~~
> In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
> ../src/solidity-frontend/solidity_convert.h:38:16: error:   'unsigned int solidity_convertert::current_scope_var_num' [-Werror=reorder]
>    38 |   unsigned int current_scope_var_num;
>       |                ^~~~~~~~~~~~~~~~~~~~~
> ../src/solidity-frontend/solidity_convert.cpp:15:1: error:   when initialized here [-Werror=reorder]
>    15 | solidity_convertert::solidity_convertert(
>       | ^~~~~~~~~~~~~~~~~~~
> cc1plus: all warnings being treated as errors
> ```

Thank you for informing that! I will resolve the errors for this PR, but the implicit constructor may be implemented in another PR because I am having exams soon.

**Comment by Sleepytt1210** [comment]
> > Thank you very much for your work. Here is my comments:
> > ```
> > - You might need to consider the implicit constructor.
> >   
> >   - In current ESBMC-Solidity, if there is no constructor function, we will manually add a default constructor by `add_implicit_constructor()`.
> > 
> > - So you might need to change the `id` format in:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```c++
> >  bool solidity_convertert::add_implicit_constructor()
> > {
> >   std::string name, id;
> >   name = id = current_contractName;     // set the idstring format 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and in any place where a function symbol is looked up by calling `context.find_symbol(id)`, e.g
> > ```c++
> > bool solidity_convertert::get_implicit_ctor_call(
> >   const int ref_decl_id,
> >   exprt &new_expr)
> > {
> >   std::string name, id;
> >   if(get_contract_name(ref_decl_id, name))
> >     return true;
> >   id = name;      // set the id format 
> >   if(context.find_symbol(id) == nullptr)
> >     return true;
> > ```
> 
> May I ask if you have a plan to implement this? It could be another PR though.
> 
> Also, you may want to resolve the build error. The PR can only be merged if all checks are passed.
> 
> ```shell
> In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
> ../src/solidity-frontend/solidity_convert.h: In constructor 'solidity_convertert::solidity_convertert(contextt&, nlohmann::json&, const string&, const string&)':
> ../src/solidity-frontend/solidity_convert.h:48:15: error: 'solidity_convertert::current_contractName' will be initialized after [-Werror=reorder]
>    48 |   std::string current_contractName;
>       |               ^~~~~~~~~~~~~~~~~~~~
> In file included from ../src/solidity-frontend/solidity_convert.cpp:1:
> ../src/solidity-frontend/solidity_convert.h:38:16: error:   'unsigned int solidity_convertert::current_scope_var_num' [-Werror=reorder]
>    38 |   unsigned int current_scope_var_num;
>       |                ^~~~~~~~~~~~~~~~~~~~~
> ../src/solidity-frontend/solidity_convert.cpp:15:1: error:   when initialized here [-Werror=reorder]
>    15 | solidity_convertert::solidity_convertert(
>       | ^~~~~~~~~~~~~~~~~~~
> cc1plus: all warnings being treated as errors
> ```

Hi,

I have a test that will not be fixed in this PR. Is there any way to mark it be ignored, or should I just remove the test case?

**Comment by lucasccordeiro** [comment]
You could use `KNOWNBUG` instead of `CORE`.

**Comment by ChenfengWei0** [comment]
@kunjsong01 Do you know how to resolve the C/C++ code-style error? 👀

**Comment by Sleepytt1210** [comment]
> @kunjsong01 Do you know how to resolve the C/C++ code-style error? 👀

I think I will have to code it like the git diff suggested. I will fix this asap.

**Comment by lucasccordeiro** [comment]
@kunjsong01 and @ChenfengWei0 : are you happy with this PR?

**Review by kunjsong01 — APPROVED** [review]
LGTM

**Comment by lucasccordeiro** [comment]
@Sleepytt1210 : could you please rebase this PR with the master branch? You could also squash some commits with the command.

**Review by ChenfengWei0 — APPROVED** [review]
LGTM. Many thanks @Sleepytt1210

**Comment by Sleepytt1210** [comment]
> @Sleepytt1210 : could you please rebase this PR with the master branch? You could also squash some commits with the command.

Yes, will do.

**Comment by lucasccordeiro** [comment]
Many thanks, @Sleepytt1210. As soon as you rebase and squash your commits, we'll merge your code into the master.

**Comment by lucasccordeiro** [comment]
@Sleepytt1210:

````
The following tests FAILED:
	832 - regression/esbmc-solidity/constructor_1 (Failed)
	834 - regression/esbmc-solidity/constructor_3 (Failed)
	835 - regression/esbmc-solidity/constructor_4 (Failed)
	836 - regression/esbmc-solidity/constructor_5 (Failed)
	837 - regression/esbmc-solidity/constructor_6 (Failed)
	881 - regression/esbmc-solidity/op_ternary_2 (Failed)
	885 - regression/esbmc-solidity/return_1 (Failed)
	886 - regression/esbmc-solidity/return_2 (Failed)
	887 - regression/esbmc-solidity/return_3 (Failed)
Errors while running CTest
````

**Comment by Sleepytt1210** [comment]
> @Sleepytt1210:
> 
> ```
> The following tests FAILED:
> 	832 - regression/esbmc-solidity/constructor_1 (Failed)
> 	834 - regression/esbmc-solidity/constructor_3 (Failed)
> 	835 - regression/esbmc-solidity/constructor_4 (Failed)
> 	836 - regression/esbmc-solidity/constructor_5 (Failed)
> 	837 - regression/esbmc-solidity/constructor_6 (Failed)
> 	881 - regression/esbmc-solidity/op_ternary_2 (Failed)
> 	885 - regression/esbmc-solidity/return_1 (Failed)
> 	886 - regression/esbmc-solidity/return_2 (Failed)
> 	887 - regression/esbmc-solidity/return_3 (Failed)
> Errors while running CTest
> ```

I think I made some errors during rebase, I will fix this.

**Comment by lucasccordeiro** [comment]
Many thanks for this PR, @Sleepytt1210.

**Comment by lucasccordeiro** [comment]
Many thanks for this PR, @Sleepytt1210.

**Comment by Sleepytt1210** [comment]
You're welcome, thank you for the opportunity too.

**Pull request merged**